addons/udes_report/models/stock_export.py

```diff
@@ -62,79 +62,103 @@ def run_stock_file_export(self, context=None, send_file_via='user'):
             quants_by_prod_by_loc[quant_prods] = get_quants_by_loc(quants_u)
 
-        def get_prod_qty(quants):
-            return sum(Quant.union(*quants).mapped('quantity'))
-
-        _logger.info(_('Creating Excel file'))
-        file_name = 'warehouse_stock_{TIMESTAMP}.xls'.format(
-            TIMESTAMP=datetime.now())
-        wb = xlwt.Workbook()
-
-        # 1/2) 'Stock File': one entry for each (prod, package) pair
+        stock_file_data = {"name": 'Stock File'}
 
-        stock_sheet = _create_sheet(wb, 'Stock File', ['Part Number',
-                                                       'Location',
-                                                       'Package',
-                                                       'Quantity'])
+        if show_tracking:
+            stock_file_data["column_titles"] = ['Part Number',
+                                                'Location',
+                                                'Package',
+                                                'Lot/Serial Number',
+                                                'Quantity']
+        else:
+            stock_file_data["column_titles"] = ['Part Number', 'Location', 'Package', 'Quantity']
 
-        row = 0
+        rows = []
         for prod, quants_by_location in quants_by_prod_by_loc.items():
             for location, quants in quants_by_location.items():
-                for pkg_name, grouped_quants in groupby(quants,
+                for pkg_name, grouped_quants in groupby(quants.sorted(key=by_package_name),
                                                         key=by_package_name):
-                    row += 1
-                    qty = get_prod_qty(grouped_quants)
-
-                    stock_sheet.write(row, 0, prod.default_code)
-                    stock_sheet.write(row, 1, location.display_name)
-                    stock_sheet.write(row, 2, pkg_name)
-                    stock_sheet.write(row, 3, qty)
-
-                    prod_summary[prod]['qty'] += qty
+                    # If lots are not present then this will not change the rows
+                    for lot_name, lot_grouped_quants in groupby(grouped_quants,
+                                                                key=by_lot_name):
+                        qty = sum(Quant.union(*lot_grouped_quants).mapped('quantity'))
+                        row = {'Part Number': prod.default_code,
+                               'Location': location.display_name,
+                               'Package': pkg_name,
+                               'Quantity': qty}
+                        if show_tracking:
+                            row['Lot/Serial Number'] = lot_name
+                        rows.append(row)
+                        prod_summary[prod]['qty'] += qty
                     if pkg_name:
                         prod_summary[prod]['n_pkgs'] += 1
+        stock_file_data['rows'] = rows
 
         # 2/2) 'Stock Summary': one entry for each prod
```

Comment:
I think the counter part to this `# 1/2 ....` is now missing.

Reply:
Good spot, added an edited version back in